ibis/backends/flink/utils.py

```diff
@@ -0,0 +1,232 @@
+import datetime
+import math
+from abc import ABC, abstractmethod
+from collections import defaultdict
+
+from ibis.common.temporal import IntervalUnit
+from ibis.util import convert_unit
+
+MIN_ALLOWED_PRECISION = {
+    IntervalUnit.YEAR: 1,
+    IntervalUnit.MONTH: 2,
+    IntervalUnit.DAY: 1,
+    IntervalUnit.HOUR: 2,
+    IntervalUnit.MINUTE: 2,
+    IntervalUnit.SECOND: 0,
+}
+
+MAX_ALLOWED_PRECISION = {
+    IntervalUnit.YEAR: 4,
+    IntervalUnit.MONTH: 2,
+    IntervalUnit.DAY: 6,
+    IntervalUnit.HOUR: 2,
+    IntervalUnit.MINUTE: 2,
+    IntervalUnit.SECOND: 9,
+}
+
+
+MICROSECONDS_IN_UNIT = {
+    unit: datetime.timedelta(**{unit.plural: 1}).total_seconds() * 10**6
+    for unit in [
+        IntervalUnit.DAY,
+        IntervalUnit.HOUR,
+        IntervalUnit.MINUTE,
+        IntervalUnit.SECOND,
+    ]
+}
+MICROSECONDS_IN_UNIT[IntervalUnit.MONTH] = (
+    30 * datetime.timedelta(days=1).total_seconds() * 10**6
+)
+MICROSECONDS_IN_UNIT[IntervalUnit.YEAR] = (
+    365 * datetime.timedelta(days=1).total_seconds() * 10**6
+)
+
+
+def _calculate_precision(interval_value: int) -> int:
+    """Calculate interval precision.
+
+    FlinkSQL interval data types use leading precision and fractional-
+    seconds precision. Because the leading precision defaults to 2, we need to
+    specify a different precision when the value exceeds 2 digits.
+
+    (see
+    https://learn.microsoft.com/en-us/sql/odbc/reference/appendixes/interval-literals)
+    """
+    # log10(interval_value) + 1 is equivalent to len(str(interval_value)), but is significantly
+    # faster and more memory-efficient
+    if interval_value == 0:
+        return 0
+    if interval_value < 0:
+        raise ValueError(
+            f"Expecting value to be a non-negative integer, got {interval_value}"
+        )
+    return int(math.log10(interval_value)) + 1
+
+
+def _format_value_with_precision(value: int, precision: int) -> str:
+    return str(value).zfill(precision)
+
+
+def _format_precision(precision: int, unit: IntervalUnit) -> str:
+    if precision > MAX_ALLOWED_PRECISION[unit]:
+        raise ValueError(
+            f"{precision} is bigger than the allowed precision for {unit} ({MAX_ALLOWED_PRECISION[unit]})"
+        )
+
+    return '' if precision <= 2 else f'({precision})'
+
+
+class FlinkIntervalType(ABC):
```

Comment:
what's ibis's standard on docstring? we should still try to add docstring for public classes and methods

Reply:
Added!